mono/mini/mini-llvm.c

```diff
@@ -7868,10 +7951,18 @@ emit_method_inner (EmitContext *ctx)
 		 * in load_method ().
 		 */
 		gboolean needs_init = ctx->cfg->got_access_count > 0;
-		if (!needs_init && mono_class_get_cctor (cfg->method->klass)) {
+		MonoMethod *cctor = NULL;
+		if (!needs_init && (cctor = mono_class_get_cctor (cfg->method->klass))) {
 			/* Needs init to run the cctor */
 			if (cfg->method->flags & METHOD_ATTRIBUTE_STATIC)
 				needs_init = TRUE;
+			if (cctor == cfg->method)
+				needs_init = FALSE;
+
+			// If we are a constructor, we need to init so the static
+			// constructor gets called.
+			if (!strcmp (cfg->method->name, ".ctor"))
+				needs_init = TRUE;
```

Comment:
This shouldn't be needed.

Reply:
It seemed like it was at some point, but this PR has undergone some major design changes since then. Testing if it works without right now.

Reply:
Disabling this makes static-ctor fail.

Reply:
With it disabled: 

```
; Function Attrs: noinline uwtable
define hidden monocc void @X__cctor() #4 gc "mono" {
BB0:
  %is_inited = load i8, i8* getelementptr inbounds ([5 x i8], [5 x i8]* @mono_inited, i64 0, i64 0), align 1
  %0 = icmp eq i8 %is_inited, 0
  %1 = load i64*, i64** getelementptr inbounds ([34 x i64*], [34 x i64*]* @mono_aot_static_ctor_llvm_got, i64 0, i64 7), align 8
  %2 = load i64, i64* %1, align 4
  %3 = icmp eq i64 %2, 0
  br i1 %3, label %gc.safepoint_poll.exit, label %gc.safepoint_poll.poll.i

gc.safepoint_poll.poll.i:                         ; preds = %BB0
  %4 = load void ()*, void ()** bitcast (i64** getelementptr inbounds ([34 x i64*], [34 x i64*]* @mono_aot_static_ctor_llvm_got, i64 0, i64 25) to void ()**), align 8
  call void %4() #6
  br label %gc.safepoint_poll.exit

gc.safepoint_poll.exit:                           ; preds = %BB0, %gc.safepoint_poll.poll.i
  br i1 %0, label %NOTINITED_BB3, label %BB2, !prof !15

BB2:                                              ; preds = %NOTINITED_BB3, %gc.safepoint_poll.exit
  %SFLDA_301 = load i32*, i32** bitcast (i64** getelementptr inbounds ([34 x i64*], [34 x i64*]* @mono_aot_static_ctor_llvm_got, i64 0, i64 30) to i32**), align 16, !invariant.load !16
  store i32 1, i32* %SFLDA_301, align 4
  ret void

NOTINITED_BB3:                                    ; preds = %gc.safepoint_poll.exit
  call monocc void @mono_aot_static_ctorinit_method(i32 0)
  br label %BB2
}

; Function Attrs: noinline norecurse nounwind readnone uwtable
define hidden monocc void @X__ctor(i64* nocapture %this) #5 gc "mono" {
BB0:
  %0 = load i64*, i64** getelementptr inbounds ([34 x i64*], [34 x i64*]* @mono_aot_static_ctor_llvm_got, i64 0, i64 7), align 8
  %1 = load i64, i64* %0, align 4
  %2 = icmp eq i64 %1, 0
  br i1 %2, label %gc.safepoint_poll.exit, label %gc.safepoint_poll.poll.i

gc.safepoint_poll.poll.i:                         ; preds = %BB0
  %3 = load void ()*, void ()** bitcast (i64** getelementptr inbounds ([34 x i64*], [34 x i64*]* @mono_aot_static_ctor_llvm_got, i64 0, i64 25) to void ()**), align 8
  call void %3() #6
  br label %gc.safepoint_poll.exit

gc.safepoint_poll.exit:                           ; preds = %BB0, %gc.safepoint_poll.poll.i
  ret void
}
``` 

And main doesn't emit any cctor references. 

When this code runs using the PLT code, the place where the static constructor gets called is when we init X's instance constructor. Preserving that behavior is most easily done by making the change mentioned here.